### DISCUSSION TITLE
Blueprints workflow

### DISCUSSION THREAD

**Discussion by perkalkyl** [nxrighthere/UnrealCLR#41] [Help] [answered]
Hi! I am totally new to Unreal, but have spent a day going through the online tutorials so I have an ok grasp of the workflow now.
As I am a C# developer, I am evaluating UnrealCLR as a way to build my logic.
As the title says, I successfully downloaded and built the UnrealCLR project from GitHub. Following the excellent "Getting Started" to the point, I can play the different tests on an empty project. So far so good.
But here is where my issues start. Looking at the "C# in Unreal Engine using UnrealCLR" video, it all makes sense up til he shows how the blueprint should look for the first sample. Mine doesn't look like that at all. I have no FindManagedFunction or ExecuteManagedFunction items in my blueprint (as hopefully is visible in the image i tried to upload).

![image](https://user-images.githubusercontent.com/27836749/94209435-a6983d80-fecc-11ea-8756-4b06ba8833b1.png)

So what am i missing here? I am running Unreal 4.25.3 and Visual Studio 2019 on a Windows 10 x64 machine.

The second question (that is really the extension of the first question) I have is: How do you debug your functions?? If I make changes to the Tests classes, the change shows up after recompile and re-install the plugin. Does that mean that there is no way to runtime-debug the classes created in C#??

I would love some pointers here, because I am super-pepped to get started on this :-)

Best regards
Per

**Comment by nxrighthere** [comment, marked as answer]
Hello and welcome.

> I have no FindManagedFunction or ExecuteManagedFunction items in my blueprint

What you are looking for is blueprint nodes that are accessible by right-clicking on any graph and selecting the .NET category on the list, then just create nodes and connect them. You are free to manipulate nodes as you want.

Most likely, these nodes will be changed in upcoming versions.

> How do you debug your functions?? If I make changes to the Tests classes, the change shows up after recompile and re-install the plugin. Does that mean that there is no way to runtime-debug the classes created in C#??

You don't need to reinstall the entire plugin after making changes to your assemblies. Just modify your C# code, recompile the .NET project to the managed folder, and hit play as [described](https://github.com/nxrighthere/UnrealCLR/blob/master/MANUAL.md#project) in the manual.

To debug managed code you have to use IDE with an attached debugger to the editor. You also can use functionality from the [`Assert`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Assert.md) and [`Debug`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Debug.md) classes.

**Reply by perkalkyl** [reply]
Hi again. I got debugging running, and I finally get the concept better :-P This is sooo funky!! Amazing job with the plugin. The fact that the VS debugger can edit code while in break mode and just keep going is fantastic. I didn't expect that to work via your plugin.

With your permission maybe I could give a few pointers on how to make the relation between the plugin and your test codes a bit more understandable for a total newbee. I have been working as a developer for longer than I can remember, and have spent my life inside the Visual studio Environment, building WinForm, ASP.NET Core etc applications in C# for the past decade or so. And the workprocess of using the command prompt to build and publish is long gone :-). So I had a really hard time before I understood that the Contents, like the tests enum, the tests blueprint etc.  that gets added to the project when using the installation project really is not a product automatically produced by the code :-) I tried to add a function to the tests project but couldn't get it to show up in the Tests Blueprint content :-D I know better now.

I'll get back to you after I've spent som more time with this.
But again, Thanx for your effort. This might be da shit for a lot of us developers that left C++ a looong time ago.

**Reply by nxrighthere** [reply]
Sure, you are welcome!

Yes, the persistent content of the engine is independent of the managed code. My goal is to make it possible to generate some primitives at runtime in later versions. Priorities are just a bit different for now.

I'm glad that you like it, enjoy!

**Comment by nxrighthere** [comment]
Also, see [this](https://github.com/nxrighthere/UnrealCLR#assemblies-management) and other sections from the overview.
